libutils/stack.c

```diff
@@ -0,0 +1,267 @@
+        return NULL;
+    }
+
+    pthread_mutexattr_destroy(&attr);
+
+    stack->capacity = initial_capacity;
+    stack->size = STACK_EMPTY;
+    stack->data = xcalloc(initial_capacity, sizeof(void *));
+    stack->ItemDestroy = ItemDestroy;
+
+    return stack;
+}
+
+void StackDestroy(Stack *stack)
+{
+    if (stack != NULL)
+    {
+        ThreadLock(stack->lock);
+        DestroyRange(stack, STACK_EMPTY, stack->size);
+        ThreadUnlock(stack->lock);
+
+        StackSoftDestroy(stack);
+    }
+}
+
+void StackSoftDestroy(Stack *stack)
+{
+    if (stack != NULL)
+    {
+        if (stack->lock != NULL)
+        {
+            pthread_mutex_destroy(stack->lock);
+            free(stack->lock);
+        }
+
+        free(stack->data);
+        free(stack);
+    }
+}
+
+void *StackPop(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+
+    size_t size = stack->size;
+    void *item = NULL;
+
+    if (size > STACK_EMPTY)
+    {
+        size--;
+        item = stack->data[size];
+
+        stack->data[size] = NULL;
+        stack->size = size;
+    }
+
+    ThreadUnlock(stack->lock);
+
+    return item;
+}
+
+void StackPush(Stack *stack, void *item)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+
+    ExpandIfNecessary(stack);
+    stack->data[stack->size++] = item;
+
+    ThreadUnlock(stack->lock);
+}
+
+size_t StackPushReportCount(Stack *stack, void *item)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+
+    ExpandIfNecessary(stack);
+    stack->data[stack->size++] = item;
+    size_t size = stack->size;
+
+    ThreadUnlock(stack->lock);
+
+    return size;
+}
+
+size_t StackCount(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+    size_t count = stack->size;
+    ThreadUnlock(stack->lock);
+
+    return count;
+}
+
+size_t StackCapacity(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+    size_t capacity = stack->capacity;
+    ThreadUnlock(stack->lock);
+
+    return capacity;
+}
+
+bool StackIsEmpty(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+    bool const empty = (stack->size == STACK_EMPTY);
+    ThreadUnlock(stack->lock);
+
+    return empty;
+}
+
+Stack *StackCopy(Stack *stack)
+{
+    assert(stack != NULL);
+
+    ThreadLock(stack->lock);
+
+    Stack *new_stack = StackNew(stack->capacity, stack->ItemDestroy);
+    for (size_t i = 0; i < stack->size; i++)
+    {
+        new_stack->data[i] = stack->data[i];
+    }
+
+    new_stack->size = stack->size;
+
+    ThreadUnlock(stack->lock);
+
+    return new_stack;
+}
+
+/**
+  @brief Destroys data in range.
+  @note Assumes that locks are acquired.
+  @param [in] stack Pointer to struct.
+  @param [in] start Start position to destroy from.
+  @param [in] end Where to stop.
+  */
+static void DestroyRange(Stack *stack, size_t start, size_t end)
+{
+    assert(stack != NULL);
+    if (start > stack->capacity || end > stack->capacity)
```

Comment:
Since start is unsigned, wouldn't that scenario be impossible anyway? It already checks for if it overflows.

Reply:
Yeah, but what if `STACK_EMPTY` becomes `1` at some point. I think we should either use the constant (macro) or get rid of it.

Reply:
This will create a warning, though. Not sure we want that just because `STACK_EMPTY` might be `1` at some point. The functionality for the stack is also pretty basic, so I don't think that would be necessary.

Reply:
Then I think it makes sense to get rid of the macro. I'd say it is more confusing than just `0` in this code.